JDK-8283326: Implement SafeFetch statically

tstuefe opened the pull request:
Hi,

This is a reimplementation of SafeFetch to get rid of dynamic code generation. This is done, depending on platform, via static assembly, Posix setjmp/longjmp or SEH.

This is based on a proposal by @theRealAph [1] when discussing [JDK-8282475](https://bugs.openjdk.java.net/browse/JDK-8282475). I extended the proposal to do this for all platforms, not just MacOS, see discussion below.

---

Why we do this:

SafeFetch is special: it is a basic utility function that may be called before the VM is properly initialized, from a "foreign" thread not attached to the VM, in a broken environment... mainly this is due to its use during stack walking, error reporting, hex dumping, etc. SafeFetch has to be safe to call under any circumstances - or as safe as we can make it.

Today, SafeFetch is implemented via stub routines. That causes problems:

- MacOs aarch64 has mechanisms to prevent execution of arbitrary non-text memory. On that platform, invoking generated code requires some awkward gymnastics [2]. These workarounds require access to Thread::current, which renders SafeFetch non-functional if Thread::current is NULL (e.g. crash reporting in "foreign" native threads or in a corrupted VM). They also cost a bit of performance on every SafeFetch invocation. Note that more OSes may implement similar execution prevention mechanics in the future.

- There is a time window during VM initialization where the stubs are not generated yet and thus SafeFetch is not available.

Both points cause hs-err files to be compromised (e.g. missing call stacks) for crashes in native threads, when the VM is corrupted, or early during VM initialization. Since SafeFetch is such a basic tool, it causes more issues than that (e.g., on AIX it is used for verifying mprotect() itself). But hs-err files being compromised by secondary errors is the most annoying effect.

A minor point is that generating stubs takes a tiny bit of startup time. Not a big issue, but anything we can generate at build time instead of at runtime is good.

---

The patch:

- Removes the SafeFetch stub routine generation for all platforms
- For Linux and MacOS, it implements SafeFetch as cpu-specific static assembler routines. These routines live in `os_cpu/<os>_<cpu>/safefetch_<os>_<cpu>.S`.
- On Windows, SafeFetch is implemented via SEH. This is beautifully simple and does not need handwritten assembly at all. See `os/windows/safefetch_windows.hpp`.
- On Zero, SafeFetch uses a Posix compatible solution using `sigsetjmp(2)`. This is portable and robust, but a bit slow compared to the other methods. Note that this solution existed before, and is well tested. I just moved the code around and massaged it a bit.
- AIX uses now the same method as Zero. AIX "downgraded" in that sense from the stub routine solution, but since SAP gave up maintenance of AIX, I don't have easy access to AIX hardware. If SafeFetch speed becomes an issue, I am sure a solution based on static assembly can be implemented on AIX too.
- The SafeFetch gtests were greatly expanded.

Tests:

- The patch was tested manually by building and running gtests on:
  - MacOS (x64, aarch64)
  - Linux (aarch64, arm, ppcle, s390, x64, x86, riscv)
  - Windows (x64 and x86, but not arm, see below)
  - AIX
  - Zero (on Linux x64).
- GHAs
- The patch ran through SAP nightlies for several nights in a row

Note:
- I have no access to Windows arm, so I could not test it. I contacted @mo-beck  to verify that the patch works there.
- I have no access to BSD. I am not sure who maintains BSD, so I have no idea who to contact. However, I believe the MacOS solution should work for BSD too. If not, BSD can simply fall back to the Posix sigjmp solution like AIX and Zero.

It would be nice to have @RealLucy, @TheRealMDoerr and @theRealAph to okay s390, ppcle and aarch64 (and possibly arm32). Each assembly file is tiny, so it should not take long. @RealFYang and @backwaterred already tested the patch on RiscV and AIX (thanks).

Thanks, Thomas

[1] https://github.com/openjdk/jdk/pull/7727#issuecomment-1064980598
[2] https://github.com/openjdk/jdk/blob/4df67426ed02f18af0757897acb28b636a317a91/src/hotspot/share/runtime/safefetch.inline.hpp#L37-L41

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8283326](https://bugs.openjdk.java.net/browse/JDK-8283326): Implement SafeFetch statically


### Reviewers
 * [David Holmes](https://openjdk.java.net/census#dholmes) (@dholmes-ora - **Reviewer**) ⚠️ Review applies to 8d1cc4ed30bc1024334ea34588c0a4febb29d983
 * [Martin Doerr](https://openjdk.java.net/census#mdoerr) (@TheRealMDoerr - **Reviewer**) ⚠️ Review applies to f9c2d9d0d4a84be3acb91362bbe681879a9d51d1
 * [Anton Kozlov](https://openjdk.java.net/census#akozlov) (@AntonKozlov - Committer) ⚠️ Review applies to 16520bc25726ea7cf72b198218e9159cf4b3d900
 * [Lutz Schmidt](https://openjdk.java.net/census#lucy) (@RealLucy - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/7865/head:pull/7865` \
`$ git checkout pull/7865`

Update a local copy of the PR: \
`$ git checkout pull/7865` \
`$ git pull https://git.openjdk.java.net/jdk pull/7865/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 7865`

View PR using the GUI difftool: \
`$ git pr show -t 7865`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/7865.diff">https://git.openjdk.java.net/jdk/pull/7865.diff</a>

</details>

dholmes-ora commented on src/hotspot/os/posix/signals_posix.cpp:
```diff
@@ -599,27 +593,30 @@ int JVM_HANDLE_XXX_SIGNAL(int sig, siginfo_t* info,
+  // instruction. In those cases, use siginfo si_addr instead.
+  address pc = NULL;
+  if (uc != NULL) {
+    if (S390_ONLY(sig == SIGILL || sig == SIGFPE) NOT_S390(false)) {
+      pc = (address)info->si_addr;
+    } else if (ZERO_ONLY(true) NOT_ZERO(false)) {
```
So on most platforms this will become:
```

if (false) {
...
} else if (false) {
...
}
```

I would expect compilers to complain about that. ??

Update: I see this code already exists so obviously the compilers don't complain.

dholmes-ora commented on src/hotspot/os/posix/vmError_posix.cpp:
```diff
@@ -76,11 +75,12 @@ static void crash_handler(int sig, siginfo_t* info, void* ucVoid) {
-  if (uc && pc && StubRoutines::is_safefetch_fault(pc)) {
-    os::Posix::ucontext_set_pc(uc, StubRoutines::continuation_for_safefetch_fault(pc));
+#if defined(SAFEFETCH_METHOD_STATIC_ASSEMBLY) || defined(SAFEFETCH_METHOD_SIGSETJMP)
+  // Handle safefetch here too, to be able to use SafeFetc() inside the error handler
```
Typo: SafeFetc()

dholmes-ora replied:
Ditto regarding whether the guard is needed.

dholmes-ora replied:
Again this code is effectively unconditional - the ifdefs are not needed.

dholmes-ora commented on src/hotspot/os/windows/safefetch_windows.hpp:
```diff
@@ -0,0 +1,51 @@
+  return v;
+}
+
+inline int SafeFetch32(const int* adr, int errValue)                { return SafeFetchXX<int>(adr, errValue); }
+inline intptr_t SafeFetchN(const intptr_t* adr, intptr_t errValue)  { return SafeFetchXX<intptr_t>(adr, errValue); }
```
Nit: these are hard to read on one long line, please use multi-lines.

dholmes-ora commented on src/hotspot/share/runtime/os.hpp:
```diff
@@ -296,6 +296,10 @@ class os: AllStatic {
   // Return the default page size.
   static int    vm_page_size();
 
+  // The largest possible standard page size (sysconf(_SC_PAGESIZE)) across all our
+  // platforms, usable as a compile-time constant.
+  static const size_t max_page_size_crossplatform = 64 * K; // ppc has 64 K pages.
+
```
What is this for??

tstuefe replied:
A remnant of an older implementation that slipped through. Thanks for catching, I'll remove it.

dholmes-ora commented on src/hotspot/share/runtime/safefetch.hpp:
```diff
@@ -0,0 +1,45 @@
+#ifdef _WIN32
+  // Windows uses Structured Exception Handling
+  #include "safefetch_windows.hpp"
+  #define SAFEFETCH_METHOD_SEH
```
This doesn't seem to be needed.

dholmes-ora commented on src/hotspot/os/posix/signals_posix.cpp:
```diff
@@ -599,27 +593,30 @@ int JVM_HANDLE_XXX_SIGNAL(int sig, siginfo_t* info,
-#endif // ZERO
   }
 
+#if defined(SAFEFETCH_METHOD_STATIC_ASSEMBLY) || defined(SAFEFETCH_METHOD_SIGSETJMP)
```
The guard seems unnecessary - all posix platforms need to handle safefetch here don't they?

dholmes-ora replied:
Sorry how is this "resolved". Every posix platform will define one of those cases so this code is effectively unconditional.

tstuefe replied:
I use "resolved" to mark my internal progress, which you may not see yet since I did not push yet.

dholmes-ora commented on src/hotspot/share/runtime/safefetch.hpp:
```diff
@@ -0,0 +1,45 @@
+  // This is slower than the other methods and uses more thread stack,
+  // but its safe and portable.
+  #include "safefetch_sigjmp.hpp"
+  #define SAFEFETCH_METHOD_SIGSETJMP
+#else
+  // All other platforms use static assembly
+  #include "safefetch_static.hpp"
+  #define SAFEFETCH_METHOD_STATIC_ASSEMBLY
```
It seems to me that we don't actually need these SAFEFETCH_METHOD_x defines.

dholmes-ora reviewed (changes requested):
Hi Thomas,

This looks really good! Definite approval in principle. I can't comment on the details of the .S files. A few nits and comments. Main is comment is that SAFEFETCH_METHOD_x defines don't seem to actually be needed ??

Thanks,
David

tstuefe commented on src/hotspot/os/posix/safefetch_sigjmp.cpp:
```diff
@@ -0,0 +1,107 @@
+#ifdef SAFEFETCH_METHOD_SIGSETJMP
+
+// For SafeFetch we need POSIX TLS and sigsetjmp/longjmp.
+// (Note: We would prefer compiler level TLS but for some reason __thread does not
```
Ah, that explains the weird errors I got. I'll change the comment.

tstuefe replied:
@fweimer-rh I'll move `pthread_setspecific` out of signal handling, thanks for the hint.

`pthread_getspecific` may not be signal safe, but it has been in use in JVM signal handlers for decades (signal handlers call `Thread::current`, which uses TLS) and it works in practice on all Posix platforms.

tstuefe commented:
> Hi Thomas,
> 
> This looks really good! Definite approval in principle. I can't comment on the details of the .S files. A few nits and comments. Main is comment is that SAFEFETCH_METHOD_x defines don't seem to actually be needed ??
> 
> Thanks, David

Thank you David!

SAFEFETCH_METHOD_xx was supposed to be a clean separation between safefetch method and platform, since there is no clear 1:1 relation and you can, and may need to, mix and match. Examples: 
- On Posix platforms you can use either static assembly or setjmp, and need to exclude one of them from compilation - either build safefetch_xx_xx.S, or build safefetch_sigjmp.cpp.
- On Windows, we use SEH, but I don't know yet if that works for Windows aarch64, and we still may need to switch to static assembly for this platform. I hope to avoid this though since MASM is a pain.

Therefore I'd like to keep the separation between SafeFetch method and platform, at least for now. Makes it easer to switch back and forth, and I find it actually cleaner.

Cheers Thomas

fweimer-rh commented on src/hotspot/os/posix/safefetch_sigjmp.cpp:
```diff
@@ -0,0 +1,107 @@
+  pthread_setspecific(g_jmpbuf_key, &jb);
+
+  // unsafe access
+  n = *adr;
```
Maybe this should use `*const_cast<volatile T*>(adr)` to force the read?

tstuefe replied:
I am not sure why this would be needed?

AntonKozlov commented on src/hotspot/share/runtime/safefetch_static.cpp:
```diff
@@ -0,0 +1,65 @@
+
+bool handle_safefetch(int sig, address pc, void* context) {
+  ucontext_t* uc = (ucontext_t*)context;
+  if ((sig == SIGSEGV || sig == SIGBUS) && uc != NULL) {
+    address pc = os::Posix::ucontext_get_pc(uc);
+    if (pc == (address)_SafeFetch32_fault) {
+      os::Posix::ucontext_set_pc(uc, (address)_SafeFetch32_continuation);
```
This posix-specific code should be probably moved to os/posix.

AntonKozlov commented on src/hotspot/share/runtime/safefetch_static.hpp:
```diff
@@ -0,0 +1,45 @@
+}
+#endif // _LP64
+
+inline bool CanUseSafeFetch32() { return true; }
+inline bool CanUseSafeFetchN()  { return true; }
```
AFAICS all CanUseSafeFetchXX always return true and there are few calls of these functions. It's worth to remove them completely then.

AntonKozlov commented on src/hotspot/os_cpu/bsd_aarch64/safefetch_bsd_aarch64.S:
```diff
@@ -0,0 +1,69 @@
+    ELF_TYPE(SafeFetch32,@function)
+SYMBOL(SafeFetch32):
+SYMBOL(_SafeFetch32_fault):
+    ldr      w1, [x0]
+SYMBOL(_SafeFetch32_continuation):
+    mov      x0, x1
+    ret
```
Minor: on RISC platforms you seems to prefer to pack exceptional and non-exceptional paths, and on x86 they are separated. If we align this with x86, we'd save an instruction in each execution of non-exceptional path, with an extra instruction in static code.
```
    ldr      w0, [x0]
    ret
SYMBOL(_SafeFetch32_continuation):
    mov      x0, x1
    ret
```

tstuefe replied:
I'm surprised this would work. Can you really specify the same register as load destination and as load source?

tstuefe replied:
Ah, it seems to work. You live and learn :)

AntonKozlov replied:
Yes, on arm32/64 it would not work if the instruction uses a register as a target and write-back, like `ldr r0, [r0, 4]!`, otherwise it's OK. AFAIR, it's also OK on ppc. 

As a reference, I've checked the output of gcc/clang, both don't use the extra register, so the argument regarding performance looks valid :) Not insisting on fixing, this is indeed minor.

```
anton@mercury:/tmp$ cat test.c
long test(long* addr) {
	return *addr;
}
anton@mercury:/tmp$ aarch64-linux-gnu-gcc --version
aarch64-linux-gnu-gcc (GCC) 11.2.0
anton@mercury:/tmp$ aarch64-linux-gnu-gcc -c -O3 -o test.o test.c
anton@mercury:/tmp$ aarch64-linux-gnu-objdump -d test.o

0000000000000000 <test>:
   0:	f9400000 	ldr	x0, [x0]
   4:	d65f03c0 	ret
```

dholmes-ora commented:
> On Posix platforms you can use either static assembly or setjmp, and need to exclude one of them from compilation - either build safefetch_xx_xx.S, or build safefetch_sigjmp.cpp.

Yes but we don't need to make this build time selectable. If a platform supports static then it will have a .S file and won't use the sigjmp file (which will be empty due to ifdef based on platform); otherwise no .S file and the sigjmp file is populated.

I would hope all Windows platforms support SEH.

tstuefe commented:
> > On Posix platforms you can use either static assembly or setjmp, and need to exclude one of them from compilation - either build safefetch_xx_xx.S, or build safefetch_sigjmp.cpp.
> 
> Yes but we don't need to make this build time selectable. If a platform supports static then it will have a .S file and won't use the sigjmp file (which will be empty due to ifdef based on platform); otherwise no .S file and the sigjmp file is populated.
> 

How would this work? You would add a "#if defined(zero) && defined(aix) &&..." to safefetch_sigjmp.cpp? 

I honestly find this less readable and grep-able than now...

dholmes-ora commented:
> How would this work? You would add a "#if defined(zero) && defined(aix) &&..." to safefetch_sigjmp.cpp?
> 
> I honestly find this less readable and grep-able than now...

I missed the use of SAFEFETCH_METHOD_SIGSETJMP to control the compilation of safefetch_sigjmp.cpp, but that is the only define that you need. Posix systems are either setjmp or static the only file that needs to know that is safefetch_sigjmp.cpp.

tstuefe commented:
> > How would this work? You would add a "#if defined(zero) && defined(aix) &&..." to safefetch_sigjmp.cpp?
> > I honestly find this less readable and grep-able than now...
> 
> I missed the use of SAFEFETCH_METHOD_SIGSETJMP to control the compilation of safefetch_sigjmp.cpp, but that is the only define that you need. Posix systems are either setjmp or static the only file that needs to know that is safefetch_sigjmp.cpp.

That, and safefetch_static.cpp. Those two files contain the two variants of handle_safefetch().

tstuefe commented:
Hi David, Florian and Anton,

thanks for reviewing this. Changes:

- In safefetch_sigjmp.cpp, moved pthread_setspecific out of signal handling. Thanks to @fweimer-rh for catching this.
- In safefetch_sigjmp.cpp, reset TLS slot before accessing output variable for writing.
- Following Antons proposal, I removed CanSafeFetchxx. I wondered for a brief moment it was worth keeping it, since strictly speaking two of the three methods need signal handlers to work, so cannot work before we install the hotspot signal handler. But that would incur runtime checks which are not really useful in practice: all situations where SafeFetch are invoked are guaranteed to happen when signal handler has been installed (e.g. hs-err files can only get written if our handler runs).
- Following Antons proposal, I changed the RISC implementations to use one operation less per standard path by loading into directly into the return register.
- I whittled the usage of SAFEFETCH_METHOD_ macros down to the bare minimum needed to control which implementation Posix platforms use.
- Moved safefetch_static.cpp to os/posix/safefetch_static_posix.cpp.

+ minor stuff, too small to mention, all review driven.

Thanks for reviewing!

Thomas

dholmes-ora reviewed (approved):
Updates look good - thanks.

I think perhaps we could simplify the ifdefs even further but I'm okay with them as they are.

Thanks,
David

tstuefe commented:
> Updates look good - thanks.
> 

Thanks a lot, David!

> I think perhaps we could simplify the ifdefs even further but I'm okay with them as they are.

Yes, we may reduce this coding even further. Let's see how this works out.

Cheers, Thomas

> 
> Thanks, David

TheRealMDoerr commented on src/hotspot/os_cpu/bsd_aarch64/safefetch_bsd_aarch64.S:
```diff
@@ -0,0 +1,71 @@
+    #  x0 : address
+    #  w1 : defaultval
+
+	# needed to align function start to 4 byte
```
indentation

tstuefe commented:
> It's very unfortunate that such a simple task requires so much platform code. Would be nice if the Windows implementation would work everywhere :-) I'm not aware of a simple implementation which works reliably on all platforms, so LGTM. 

Thanks Martin!

Honestly, the new implementation does not add much complexity which has not been there before. Discounting test addition, we have about 150 locs more, most caused by the unfortunate code duplication between bsd and Linux. But as you said, there is no easy separation here. I have fought with this aspect of the patch for too long now, and I think this is the cleanest and least surprising solution. It also follows what the copy_bytes_ assembly routines did before.

> Would it make sense to add a test case which uses a negative int as default value for SafeFetch32? Just to make sure the signed int gets passed correctly through the assembler stubs and reaches C++ code without loss of sign extension.

I fixed the indentations and extended the gtest as you proposed. I re-ran tests on x64, x86, arm-32 and aarch64, looks good.

Thanks!

Cheers Thomas

TheRealMDoerr reviewed (approved):
Thanks for the update! Thumbs up from my side!

AntonKozlov reviewed (commented):
That's Indeed unfortunate, although reasonable, that asm files cannot be placed cpu-specific directory. A comment below.

tstuefe commented on src/hotspot/share/runtime/safefetch.hpp:
```diff
@@ -0,0 +1,46 @@
+#define SHARE_RUNTIME_SAFEFETCH_HPP
+
+#include "utilities/macros.hpp"
+
```
Yes, this bugged me too. I tried to keep the windows implementations inline. Your proposal makes sense, I'll do that.

tstuefe commented:
I wrapped the functions into inline wrappers, which live in safefetch.hpp - easy to see that way. 

Re-tested on linux x86, 64, arm, aarch, zero, ppcle, windows x64, x86, mac m1.

I hope this is the last major change, re-testing all this stuff is a lot of work...

AntonKozlov reviewed (approved):
Really appreciate the last change. A last minor comment is below. In the rest, looks good. Thank you!

tstuefe commented on src/hotspot/share/runtime/safefetch.hpp:
```diff
@@ -0,0 +1,55 @@
+#endif
+
+
+inline int SafeFetch32(int* adr, int errValue) {
```
Thank you, Anton. I re-added the comment as you proposed.

RealLucy reviewed (approved):
Changes look good to me.

tstuefe commented:
> Changes look good to me.

Thanks @RealLucy! Nice to have s390 expert approval :)

tstuefe commented:
Happily, Microsoft reported back that my patch works on Windows aarch64 too. Since all other platforms are fine, I'll push.

Thanks to @theRealAph for the idea, @fweimer-rh for the initial help, @dholmes-ora , @AntonKozlov , @TheRealMDoerr and @RealLucy for the reviews. This has been a fun little project.

/integrate
